### DISCUSSION TITLE
Delegate OAuth2PasswordRequestForm as sub-dependency

### DISCUSSION THREAD

**Discussion by Rfferrao87** [fastapi/fastapi#7045] [Questions] [answered]
I'm trying to implement OAuth2 authentication for my `fastapi-0.61.2` REST API, but it doesn't seem to work with the following setup:

- **token.py**
```
user_auth = authentication.UserAuthenticator()

@router.post('/', response_model=schemas.Token)
async def get_token(user: schemas.UserName = Depends(user_auth)):
...
```

- **authentication.py**
```
class UserAuthenticator:
...
    def __call__(self, form_data: OAuth2PasswordRequestForm = Depends()):
...
```

With this, I was expecting to have access to the form respective to `OAuth2PasswordRequestForm` in FastAPI's `openapi.json` page, but the request body is empty.

I have tried referencing it directly in the `token` endpoint and it came up. For reference:

```
@router.post('/', response_model=schemas.Token)
async def get_token(form_data: OAuth2PasswordRequestForm = Depends()):
...
```

[![enter image description here][1]][1]

Even though it kind of works like this, I would really like to make the initial configuration work in order to achieve my envisioned structure. Any ideas?

  [1]: https://i.stack.imgur.com/9V0Tx.png

**Comment by Mause** [comment]
You're missing the `self` argument in your method

**Comment by Rfferrao87** [comment]
Thanks, missed updating that part; still doesn't work as intended - that is, by what is described in the question - though.

**Comment by Mause** [comment, marked as answer]
Really? It seems to work for me:

```python

from fastapi.testclient import TestClient
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from fastapi import FastAPI, Depends

router = FastAPI()


class UserAuthenticator:
    def __call__(self, form_data: OAuth2PasswordRequestForm = Depends()):
        ...


user_auth = UserAuthenticator()


@router.post('/')
async def get_token(user = Depends(user_auth)):
    ...


tc = TestClient(router)

print('grant_type' in tc.get('/openapi.json').text)
```

**Comment by Rfferrao87** [comment]
Oh wow, sorry about that. I was almost 100% sure I tested it again earlier, but, as you demonstrated, it works fine after the correction. Thank you for your time!

**Comment by tiangolo** [comment]
Thanks for the help here @Mause ! :clap: :bow: 

Thanks for reporting back and closing the issue @Rfferrao87  :+1:
